### PR TITLE
Eliminate redundant subexpressions

### PR DESCRIPTION
**Pull request by MaximKalininMS**
**Description**: Apply local value numbering to merge graph nodes that will always
evaluate to the same value.

**Motivation and Context**
When a model is exported from a high-level framework, and the original source code reuses some functions, these functions may create multiple graph portions that compute the same expression. This change eliminates redundant computation.

As an alternative, this optimization could be applied on case-by-case basis in the original source code (e.g. PyTorch). However, existence of redundant computation may be non-obvious; also, if part of the source code is third-party, it complicates things even more.

This implementation is quite basic and could be improved in future. For example, we could merge identical constant initializers or apply global value numbering (working across subgraphs).

### CODE REVIEW

**Comment by MaximKalininMS** [comment]
The implementation probably needs improvement (at least more tests). Please let me know if I'm on the right track.

As an example, a recent model had 30% gains from manual CSE in pytorch. This change doesn't yet improve that specific model, but with other possible optimizations (e.g. more intelligent shape inference), it looks promising.

**Comment by MaximKalininMS** [comment]
I fixed a few issues and added tests.

What is the best way to handle non-deterministic operators with no seed? Since currently the seed is resolved at kernel-build time, rather than in the graph, these operators will look identical in the graph, but later become different in the kernel. Right now I'm just hard-coding the black list of these operators. White list would probably be more robust, but the best way IMHO is to resolve the seed at load time. Let me know what's the preferred way.

**Comment by skottmckay** [comment]
Can you provide more of a description of how it's intended to work? I roughly understand what it's trying to do, but it would be easier to review with a bit more info on how it goes about that.

File: .editorconfig [hunk 1]
```diff
@@ -0,0 +1,4 @@
+[*]
+indent_style = space
+indent_size = 2
+
```
Comment (skottmckay) [inline .editorconfig]: Adding this potentially creates confusion about what is expected in terms of formatting, and the coding guidelines call for clang-format (for which there's a .clang-format file in the root directory).

Please apply the clang-formatting to your changes.

File: onnxruntime/core/optimizer/common_subexpression_elimination.h [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+namespace onnxruntime {
+
+/**
+@Class CommonSubexpressionElimination
+Merge nodes that always evaluate to the same result.
+*/
+class CommonSubexpressionElimination : public GraphTransformer {
+ public:
+  CommonSubexpressionElimination(const std::unordered_set<std::string>& compatible_execution_providers = {}) noexcept
+      : GraphTransformer("AttentionFusion", compatible_execution_providers) {}
```
Comment (skottmckay) [inline onnxruntime/core/optimizer/common_subexpression_elimination.h]: >AttentionFusion [](start = 26, length = 15)

Update name to match the optimizer name

File: onnxruntime/core/optimizer/common_subexpression_elimination.cc [hunk 3]
```diff
@@ -0,0 +1,408 @@
...
+      , hash_(CalculateHash())
+    { }
+
+  private:
+    std::size_t CalculateHash() const;
+
+    // Operation and domain of the node that produces this value.
+    const std::string op_type_;
+    const std::string domain_;
+
```
Comment (skottmckay) [inline onnxruntime/core/optimizer/common_subexpression_elimination.cc]: could be references

**Comment by MaximKalininMS** [comment]
> 
> 
> Can you provide more of a description of how it's intended to work? I roughly understand what it's trying to do, but it would be easier to review with a bit more info on how it goes about that.

This PR basically implements [this](https://en.wikipedia.org/wiki/Value_numbering#Local_value_numbering). There are two stages: the first one assigns a value number (represented by `EquivalenceClass`) for each NodeArg; the second one replaces all references to NodeArg's with the same value number with references to one selected NodeArg that has that value number. Let me know if you want me to expand on this or add a comment in source code.

File: onnxruntime/core/optimizer/common_subexpression_elimination.cc (hunk 3 above)
Comment (MaximKalininMS) [reply onnxruntime/core/optimizer/common_subexpression_elimination.cc]: These values are set to empty strings in one of the constructors (when the equivalence class represents an input or constant initializer). So they can't be references.

**Comment by MaximKalininMS** [comment]
@skottmckay Turns out I misunderstood the seed attribute; even when it's present, the operator is non-deterministic. The questions remains though: what's the best way to handle non-deterministic operators? I think it's best to specify this at the point where the operator is registered. This way we can handle custom operators too.

File: include/onnxruntime/core/graph/constants.h [hunk 4]
```diff
@@ -33,4 +33,6 @@ constexpr const char* kTensorrtExecutionProvider = "TensorrtExecutionProvider";
 constexpr const char* kNnapiExecutionProvider = "NnapiExecutionProvider";
 constexpr const char* kDmlExecutionProvider = "DmlExecutionProvider";
 constexpr const char* kAclExecutionProvider = "ACLExecutionProvider";
+const std::unordered_set<std::string> kNonDeterministicOps =
+      {"RandomUniform", "RandomNormal", "RandomUniformLike", "RandomNormalLike", "Multinomial"};
 }  // namespace onnxruntime
```
Comment (skottmckay) [inline include/onnxruntime/core/graph/constants.h:38]: Would prefer to keep this list in core/optimizer for now if possible.

File: onnxruntime/test/testdata/transform/cse/cse1.onnx [hunk 5]
```diff
@@ -0,0 +1,29 @@
+makalini:�
+
```
Comment (skottmckay) [inline onnxruntime/test/testdata/transform/cse/cse1.onnx:2]: Is it possible to add files that generate these onnx models so they're easily updated in the future? e.g. a lot of recent additions to testdata have .py files to create the model.

File: onnxruntime/core/optimizer/common_subexpression_elimination.cc [hunk 6]
```diff
@@ -0,0 +1,428 @@
...
+#include "core/graph/graph_utils.h"
+
+#include <type_traits>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+namespace onnxruntime {
+
+namespace {
```
Comment (skottmckay) [inline onnxruntime/core/optimizer/common_subexpression_elimination.cc:14]: A big comment with an overview of what is being done and how, and example/s would be really helpful here. 
Including a link to something like [https://en.wikipedia.org/wiki/Common_subexpression_elimination](https://en.wikipedia.org/wiki/Common_subexpression_elimination) would be good as well.

File: onnxruntime/core/optimizer/common_subexpression_elimination.cc [hunk 7]
```diff
@@ -0,0 +1,428 @@
...
+  explicit EquivalenceClass(const NodeArg* non_op_value)
+    : attributes_(nullptr),
+      output_index_(kInvalidOutputIndex),
+      non_op_value_(Normalize(non_op_value)),
+      discriminator_(0),
+      hash_(CalculateHash()) {
+  }
+
+  EquivalenceClass(const Node* node, const std::vector<const EquivalenceClass*>& explicit_inputs,
+                   OutputIndex output_index, int discriminator)
```
Comment (skottmckay) [inline onnxruntime/core/optimizer/common_subexpression_elimination.cc:80]: const Node& as you require it to be non nullptr

File: onnxruntime/core/optimizer/common_subexpression_elimination.cc [hunk 8]
```diff
@@ -0,0 +1,428 @@
...
+      break;
+    case onnx::AttributeProto_AttributeType_TENSOR:
+    case onnx::AttributeProto_AttributeType_GRAPH:
+    case onnx::AttributeProto_AttributeType_SPARSE_TENSOR:
+    case onnx::AttributeProto_AttributeType_TENSORS:
+    case onnx::AttributeProto_AttributeType_GRAPHS:
+    case onnx::AttributeProto_AttributeType_SPARSE_TENSORS:
+    case onnx::AttributeProto_AttributeType_UNDEFINED:
+      break;
+  }
```
Comment (skottmckay) [inline onnxruntime/core/optimizer/common_subexpression_elimination.cc:215]: Should this just break or error out?

File: onnxruntime/core/optimizer/common_subexpression_elimination.cc [hunk 9]
```diff
@@ -0,0 +1,428 @@
...
+
+  int unique_discriminator = 1;
+
+  for (NodeIndex node_index : node_topology_list) {
+    const Node* node = graph_viewer.GetNode(node_index);
+    if (node == nullptr)
+      continue;
+
+    ORT_RETURN_IF_ERROR(Recurse(*graph.GetNode(node_index), modified, graph_level, logger));
+
```
Comment (skottmckay) [inline onnxruntime/core/optimizer/common_subexpression_elimination.cc:343]: *graph.GetNode(node_index) -> *node

File: onnxruntime/core/optimizer/common_subexpression_elimination.cc [hunk 10]
```diff
@@ -0,0 +1,428 @@
...
+      input_values.push_back(it->second);
+    }
+
+    int discriminator = 0;
+    if (!IsNodeSupported(*node)) {
+      discriminator = ++unique_discriminator;
+    }
+
+    for (OutputIndex output_index = 0; output_index < static_cast<int>(node->OutputDefs().size()); ++output_index) {
+      const NodeArg* output_def = node->OutputDefs()[output_index];
```
Comment (skottmckay) [inline onnxruntime/core/optimizer/common_subexpression_elimination.cc:367]: nit: prefer calling size() once if it's not going to change across iterations. Not a big deal here but a good general approach to take.

e.g.

for (OutputIndex output_index = 0, end = static_cast<OutputIndex>(node->OutputDefs().size()); output_index < end; ++output_index) {

File: onnxruntime/core/optimizer/common_subexpression_elimination.cc [hunk 11]
```diff
@@ -0,0 +1,428 @@
...
+      if (representative.node_arg == output_def) {
+        // output_def is the representative of its equivalence class. All other values in the same class
+        // will be replaced with output_def, but output_def itself should remain.
+        continue;
+      }
+
+      if (graph_outputs.count(output_def) > 0) {
+        // Currently, eliminating a value that is the graph's output is not supported.
+        LOGS(logger, INFO) << "Not eliminating output " << output_def->Name() << " of node " << node->Name() <<
+          "[" << node->OpType() << "] because it's the graph's output.";
```
Comment (skottmckay) [inline onnxruntime/core/optimizer/common_subexpression_elimination.cc:405]: nit: INFO level may be a bit high for this message

File: onnxruntime/core/optimizer/common_subexpression_elimination.cc [hunk 12]
```diff
@@ -0,0 +1,428 @@
...
+
+struct NodeArgPtrEquality {
+  bool operator()(const NodeArg* lhs, const NodeArg* rhs) const {
+    return Normalize(lhs) == Normalize(rhs);
+  }
+};
+
+bool IsNodeSupported(const Node& node) {
+  if (node.Domain() != kOnnxDomain || node.ContainsSubgraph()) {
+    return false;
```
Comment (skottmckay) [inline onnxruntime/core/optimizer/common_subexpression_elimination.cc:295]: What's the reason for excluding other domains?

File: onnxruntime/core/optimizer/common_subexpression_elimination.cc [hunk 13]
```diff
@@ -0,0 +1,428 @@
...
+
+    for (OutputIndex output_index = 0; output_index < static_cast<int>(node->OutputDefs().size()); ++output_index) {
+      const NodeArg* output_def = node->OutputDefs()[output_index];
+      auto equivalence_class = onnxruntime::make_unique<EquivalenceClass>(node, input_values, output_index, discriminator);
+      auto* raw_ptr = equivalence_class.get();
+
+      auto it = value_to_representative.find(raw_ptr);
+      if (it == value_to_representative.end()) {
+        unique_equivalence_classes.emplace_back(std::move(equivalence_class));
+        it = value_to_representative.emplace_hint(it, raw_ptr,
```
Comment (skottmckay) [inline onnxruntime/core/optimizer/common_subexpression_elimination.cc:374]: nit: prefer push_back. [https://abseil.io/tips/112](https://abseil.io/tips/112)

File: onnxruntime/core/optimizer/common_subexpression_elimination.cc [hunk 14]
```diff
@@ -0,0 +1,428 @@
...
+
+namespace onnxruntime {
+
+Status CommonSubexpressionElimination::ApplyImpl(Graph& graph, bool& modified, int graph_level, const logging::Logger& logger) const {
+  GraphViewer graph_viewer(graph);
+  const auto& node_topology_list = graph_viewer.GetNodesInTopologicalOrder();
+
+  // Pool of equivalence classes; unique_ptr to guarantee stable address.
+  std::vector<std::unique_ptr<EquivalenceClass>> unique_equivalence_classes;
+
```
Comment (skottmckay) [inline onnxruntime/core/optimizer/common_subexpression_elimination.cc:325]: nit: could do a reserve for a rough guess of the number of entries you'll end up with. graph.NumberOfNodes() will give the number of nodes as a starting point.

File: onnxruntime/core/optimizer/common_subexpression_elimination.cc (hunk 8 above)
Comment (MaximKalininMS) [reply onnxruntime/core/optimizer/common_subexpression_elimination.cc:215]: Same hash for different values is expected (consider that there is a finite number of hashes but infinite number of attributes). The hash map will do collision resolution.

File: onnxruntime/core/optimizer/common_subexpression_elimination.cc (hunk 12 above)
Comment (MaximKalininMS) [reply onnxruntime/core/optimizer/common_subexpression_elimination.cc:295]: We don't know whether those operations are deterministic or not (e.g. custom ops)

Comment (skottmckay) [reply onnxruntime/core/optimizer/common_subexpression_elimination.cc:318]: A comment saying that will help future developers looking at this and wondering why. 

Can we support the MS Domain given those ops are part of this codebase if contrib ops are enabled?

---
In reply to: [437070556](https://github.com/microsoft/onnxruntime/pull/3047#discussion_r437070556) [](ancestors = 437070556)

Comment (skottmckay) [reply onnxruntime/core/optimizer/common_subexpression_elimination.cc:318]: As far as I know they are. I took a look and didn't notice any that weren't. Maybe just capture in a comment that these domains could potentially be enabled but leave as disabled for now. That will provide a hint if anyone is looking at this code and wondering why it doesn't work for those ops and they can carefully consider adding some/all at that point.

**Comment by MaximKalininMS** [comment]
"Linux GPU CI Pipeline" failed due to an agent error; not sure about the rest, but the crash is the same as for the two latest commits in master.

**Comment by hariharans29** [comment]
Looks like CentOS build isn't happy - 

/onnxruntime_src/onnxruntime/core/optimizer/common_subexpression_elimination.cc:254:1: error: control reaches end of non-void function [-Werror=return-type]
 }
 ^
/onnxruntime_src/onnxruntime/core/optimizer/common_subexpression_elimination.cc: In member function 'std::size_t onnxruntime::{anonymous}::DeepPointerHash::operator()(const Ptr&) const [with Ptr = const onnxruntime::{anonymous}::EquivalenceClass*; std::size_t = long unsigned int]':
/onnxruntime_src/onnxruntime/core/optimizer/common_subexpression_elimination.cc:56:3: error: control reaches end of non-void function [-Werror=return-type]
   }
   ^

**Comment by hariharans29** [comment]
@MaximKalininMS @skottmckay  - is this change good to go ?

**Comment by skottmckay** [comment]
> @MaximKalininMS @skottmckay - is this change good to go ?

Good with me.

**Pull request merged**